### ISSUE TITLE
Multiprocessping-distributed ERROR

### ISSUE THREAD

**Issue by LittleLampChen** [pytorch/pytorch#19559]
## To Reproduce

I was training Res Net on Ubuntu 16.04 with 4 TITAN XP. 
I use multiprocessing-distributed training mode. 
It works all well after many times training. But, It is very strange that while training the model, i was editing the python script, and made a careless Syntax error. Then, I saved this file with syntax error, BUT!!! the running program has crashed for this syntax error. At whole process, I haven't done anything to the running program, does the distribution training will reload the python script at every training time or any other cause?

Pythorch 1.0.0 python 3.7

**Comment by mrshenli** [comment]
> does the distribution training will reload the python script at every training time or any other cause?

Do you mean will it reload at every training iteration?  IIUC, PyTorch distributed does not change Python's module loading behavior. From what I can see in my dev env, it will not reload. But, I have two guesses for your observation. 

1. Maybe the program is not reloaded, but the error msg (might be caused by other reasons) will point back to the edited source file, which could be where the syntax error is. 
2. Maybe some process starts after you editing the source file, and hence picked up the edited version.

**Comment by soumith** [comment]
hoping to see if Shen's reply has clarified the situation. If it has not, please reopen issue with a reproducing script

**Comment by LittleLampChen** [comment]
@mrshenli @soumith Thanks for your attention. 
Recently, this has happened many times,  and mrshenli's answer doesn't work.
In order to have a better understand of this question, I list more details about my environments as following:

- RUNNING CODE
I am just using the code from https://github.com/pytorch/examples/tree/master/imagenet to train my model. 

- RUNNING PROCESSING

1. I run this code with 3 computer nodes, which i tag as CA, CB, CC.
CA, CB has totally the same configuration. Which is:

- [ Ubuntu 16.04.6 LTS (GNU/Linux 4.4.0-145-generic x86_64)] 

- [Python 3.6.8 |Anaconda, Inc.| (default, Dec 30 2018, 01:22:34) 
[GCC 7.3.0] on linux]

- [>>> import torch
>>> print(torch.__version__)
1.0.1
 ] 

- [ +-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.56       Driver Version: 418.56       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  TITAN Xp            Off  | 00000000:19:00.0 Off |                  N/A |
| 22%   40C    P0    59W / 250W |      0MiB / 12196MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  TITAN Xp            Off  | 00000000:1A:00.0 Off |                  N/A |
| 27%   45C    P0    60W / 250W |      0MiB / 12196MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  TITAN Xp            Off  | 00000000:67:00.0 Off |                  N/A |
| 30%   47C    P0    56W / 250W |      0MiB / 12196MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  TITAN Xp            Off  | 00000000:68:00.0 Off |                  N/A |
| 29%   45C    P0    56W / 250W |      0MiB / 12192MiB |      6%      Default |
+-------------------------------+----------------------+----------------------+
] 

- [ nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
] 

I was working on CC, CC is may laptop, which has the same system version with CA and CB but without GPU card.

2. In order to better coding and debug, I have mount my working directory in CC to CA and CB using nfs function. In CA, I use **VS code** as develop tool.

3. I can run this program with CA, CB together without any error. In other word, I can train model with 8 TITAN XP Card, distributed in 2 computer nodes. 

4. I can run this program with CA, CB at the same time with different tasks without any error. In other word, in CA, I train with 4 TITAN XP with batch size = 128, and in CB, I train with 4 TITAN XP with batch size = 64. They can do well at the same time, which sharing the same source code folder. (I just start CA first. then, change batch_size = 64, and start CB.)

5. It is confused that, while CA, CB training, I can edit the source code in CC safely. But, when I make an syntax error with VS-CODE in CC, Problems comes.
Most recently phenomenon is that, CA crashed with the syntax error information, but CB works well.


This seems to happened randomly, but I have meet it many times.

Thanks.

**Comment by mrshenli** [comment]
@LittleLampChen thanks for sharing the details. 

1. You mentioned that you didn't encounter this problem until recently. Did you recently switch to a different version of PyTorch? Could you please share which PyTorch version works and which breaks? BTW, it seems like people also hit this problem with Python multiprocessing ([link](https://stackoverflow.com/questions/5296977/what-will-happen-if-i-modify-a-python-script-while-its-running)).

2. what are the args that you fed into `main.py`?

**Comment by LittleLampChen** [comment]
I just use this command for one node and multi-nodes training.

This problem was occurred in pytorch 1.0.1
And I am very worried about the previous training result, if multi-process training will be influenced by source code, ~_~

**Comment by LittleLampChen** [comment]
And I use the following args

### Single node, multiple GPUs:

```bash
python main.py -a resnet50 --lr 0.01 --dist-url 'tcp://127.0.0.1:123456' --dist-backend 'nccl' --multiprocessing-distributed --world-size 1 --rank 0 [imagenet-folder with train and val folders]
```

### Multiple nodes:

Node 0:
```bash
python main.py -a resnet50 --lr 0.01 --dist-url 'tcp://192.168.0.111:123456' --dist-backend 'nccl' --multiprocessing-distributed --world-size 2 --rank 0 [imagenet-folder with train and val folders]
```

Node 1:
```bash
python main.py -a resnet50 --lr 0.01 --dist-url 'tcp://192.168.0.111:123456' --dist-backend 'nccl' --multiprocessing-distributed --world-size 2 --rank 1 [imagenet-folder with train and val folders]
```

The link has described a very similar phenomenon with mine.

**Comment by densechen** [comment]
After almost one years later, now, I know why.
If we are using multi-GPU to train our model, it seens that we have to star multi-thread for different GPUs. And each thread has to rerun the script. If we change the code during this time, the other threads may load the modified code, which caused this problem.

**Comment by JiaRu2016** [comment]
> After almost one years later, now, I know why.
> If we are using multi-GPU to train our model, it seens that we have to star multi-thread for different GPUs. And each thread has to rerun the script. If we change the code during this time, the other threads may load the modified code, which caused this problem.

@densechen 
Hi, I recently encounter same problem as yours, Could you please share more details of why this happens, and how you figure out the causation?  Any related documents or references will help, thanks~!

I'v also tired to reproduce this problem in small script, but can not, as you say, it looks really random.

**Comment by densechen** [comment]
@JiaRu2016 I think the most simple way to solve this is that do not modified your code while training.
